examples/states_ui/game.rs

```diff
@@ -0,0 +1,129 @@
+use crate::pause::PauseMenuState;
+use crate::util::delete_hierarchy;
+use amethyst::{
+    audio::output::init_output,
+    core::Time,
+    ecs::prelude::{Entity, WorldExt},
+    input::{is_close_requested, is_key_down},
+    prelude::*,
+    ui::{UiCreator, UiFinder, UiText},
+    utils::fps_counter::FpsCounter,
+    winit::VirtualKeyCode,
+};
+
+/// Main 'Game' state. Actually, it is mostly similar to the ui/main.rs content-wise.
+/// The main differences include the added 'paused' field in the state, which is toggled when
+/// 'pausing'.
+
+#[derive(Default)]
+pub struct Game {
+    paused: bool,
+    ui_root: Option<Entity>,
+    fps_display: Option<Entity>,
+    random_text: Option<Entity>,
+}
+
+impl SimpleState for Game {
+    fn on_start(&mut self, data: StateData<'_, GameData<'_, '_>>) {
+        let StateData { mut world, .. } = data;
+
+        // needed for registering audio output.
+        init_output(&mut world);
+
+        self.ui_root =
+            Some(world.exec(|mut creator: UiCreator<'_>| creator.create("ui/example.ron", ())));
+    }
+
+    fn on_pause(&mut self, _data: StateData<'_, GameData<'_, '_>>) {
+        self.paused = true;
+    }
+
+    fn on_resume(&mut self, _data: StateData<'_, GameData<'_, '_>>) {
+        self.paused = false;
+    }
+
+    fn on_stop(&mut self, data: StateData<'_, GameData<'_, '_>>) {
+        if let Some(entity) = self.ui_root {
+            delete_hierarchy(entity, data.world).expect("Failed to remove Game Screen");
+        }
+        self.ui_root = None;
+        self.fps_display = None;
+        self.random_text = None;
+    }
+
+    fn handle_event(
+        &mut self,
+        _: StateData<'_, GameData<'_, '_>>,
+        event: StateEvent,
+    ) -> SimpleTrans {
+        match &event {
+            StateEvent::Window(event) => {
+                if is_close_requested(&event) || is_key_down(&event, VirtualKeyCode::Escape) {
```

Comment:
The `is_close_requested` should likely return a `Trans::Quit` (it's when the :x: is pressed on a window, as well as alt + f4 I think). Relevant in the credits, game, menu, and paused states.

(realize it's an example, but this part could confuse myself if I were a newcomer)

Reply:
Ah, I got confused by this. In my case it still behaves as it should (since my wm does not have a :x: -button), and when trying to close it, it closes no matter what the state. But, as it turns out, my wm just kills it in these instances. Yes, I will change that